File: src/backend/utils/VIGENERE_CIPHER/service.py
# Function for key length and text length are equal
def extend_key_vigenere(key, text):
  extended_key = key
  while len(extended_key) < len(text):
      extended_key += key
  return extended_key[:len(text)]

# Encrypt text using Vigenere Cipher
def encrypt_vigenere(text, key):
  encrypted_text = ""
  for i in range(len(text)):
    encrypted_text += chr((ord(text[i]) + ord(key[i])) % 26 + 65)
  return encrypted_text

# Decrypt text using Vigenere Cipher
def decrypt_vigenere(text, key):
  decrypted_text = ""
  for i in range(len(text)):
    decrypted_text += chr((ord(text[i]) - ord(key[i]) + 26) % 26 + 65)
  return decrypted_text

File: src/backend_rails/app/utils/VIGENERE_CIPHER/service.rb
class VigenereService
  def extend_key_vigenere(key, text)
    extended_key = key
    while extended_key.length < text.length
      extended_key += key
    end
    extended_key[0, text.length]
  end

  def encrypt_vigenere(text, key)
    encrypted_text = ""
    text.each_char.with_index do |char, i|
      encrypted_text += ((char.ord + key[i].ord - 130) % 26 + 65).chr
    end
    encrypted_text
  end

  def decrypt_vigenere(text, key)
    decrypted_text = ""
    text.each_char.with_index do |char, i|
      decrypted_text += ((char.ord - key[i].ord + 26) % 26 + 65).chr
    end
    decrypted_text
  end
end